// comrade-builder-config/webpack/create-optimization.js
const path = require('path')
const glob = require('glob')

const CssMinimizerPlugin = require('css-minimizer-webpack-plugin')
const TerserPlugin = require('terser-webpack-plugin')

const cacheGroups = (__base) => {
  const result = {
    default: false
  }
  let components = glob.sync(
    path.resolve(__base + '/src/components/**/*.component.scss')
  )
  components.forEach(function (file) {
    let base = path.basename(file, '.component.scss')
    if (base !== 'header' && base !== 'footer') {
      result[`${base}`] = {
        name: base,
        filename: 'assets/js/core/[name].js',
        test: (m) => {
          const reg = new RegExp(`components/${base}$`)
          return m.constructor.name === 'CssModule' && m.context.match(reg)
        },
        chunks: 'all',
        enforce: true
      }
    }
  })
  return result
}

module.exports = function (env, __base) {
  if (env === 'development') {
    return {
      runtimeChunk: false
    }
  }
  if (env === 'production') {
    return {
      runtimeChunk: false,
      splitChunks: {
        cacheGroups: cacheGroups(__base)
      },
      minimize: true,
      minimizer: [
        new CssMinimizerPlugin(),
        new TerserPlugin({
          parallel: true
        })
      ]
    }
  }
}
